Exception on every startup with Flow 8 rc 1

I see this on every startup of the Start project using 21.0.0.rc1

```
Mon Aug 23 14:55:05 EEST 2021
There was an unexpected error (type=Internal Server Error, status=500).
java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.VaadinService.runPendingAccessTasks(com.vaadin.flow.server.VaadinSession)" because the return value of "com.vaadin.flow.server.VaadinSession.getService()" is null
com.vaadin.flow.server.ServiceException: java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.VaadinService.runPendingAccessTasks(com.vaadin.flow.server.VaadinSession)" because the return value of "com.vaadin.flow.server.VaadinSession.getService()" is null
	at com.vaadin.flow.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1594)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1557)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:302)
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:109)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:710)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:457)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:352)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:312)
	at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:141)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:52)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1060)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:962)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
```

To reproduce:
```
vaadin init hello
cd hello
mvn
# Wait for app to start, then ctrl-c
mvn
```

There seems to be two cases that can happen:

Case 1.
When pressing ctl-c, the following is logged
```
WARN 87674 --- [ionShutdownHook] com.vaadin.flow.server.VaadinSession     : A VaadinSession instance not associated to any service is getting unbound. Session destroy events will not be fired and UIs in the session will not get detached. This might happen if a session is deserialized but never used before it expires.
```
This seems to lead to the following logging on the next startup
```
2021-08-23 15:21:47.630 ERROR 87903 --- [  restartedMain] o.a.catalina.session.StandardManager     : Exception loading sessions from persistent storage

java.io.EOFException: null
	at java.base/java.io.ObjectInputStream$BlockDataInputStream.readBoolean(ObjectInputStream.java:3319) ~[na:na]
	at java.base/java.io.ObjectInputStream.readBoolean(ObjectInputStream.java:1047) ~[na:na]
	at com.vaadin.flow.server.VaadinSession.readObject(VaadinSession.java:1048) ~[flow-server-8.0.0.rc1.jar:8.0.0.rc1]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:567) ~[na:na]
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1224) ~[na:na]
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2378) ~[na:na]
```


Case 2.
Nothing is logged when pressing ctrl-c and the message in the issue is shown in the browser and logged as the first message (instead of the message about waiting for webpack to start)
```
Caused by: java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.VaadinService.runPendingAccessTasks(com.vaadin.flow.server.VaadinSession)" because the return value of "com.vaadin.flow.server.VaadinSession.getService()" is null
	at com.vaadin.flow.server.VaadinSession.unlock(VaadinSession.java:703)
	at com.vaadin.flow.server.VaadinSession.valueUnbound(VaadinSession.java:221)
	at org.apache.catalina.session.StandardSession.removeAttributeInternal(StandardSession.java:1860)
	at org.apache.catalina.session.StandardSession.removeAttribute(StandardSession.java:1337)
	at org.apache.catalina.session.StandardSession.removeAttribute(StandardSession.java:1308)
	at org.apache.catalina.session.StandardSessionFacade.removeAttribute(StandardSessionFacade.java:153)
	at com.vaadin.flow.server.WrappedHttpSession.removeAttribute(WrappedHttpSession.java:126)
	at com.vaadin.flow.server.VaadinService.readFromHttpSession(VaadinService.java:2179)
	at com.vaadin.flow.server.VaadinService.loadSession(VaadinService.java:2153)
	at com.vaadin.flow.server.VaadinService.getExistingSession(VaadinService.java:1007)
	at com.vaadin.flow.server.VaadinService.doFindOrCreateVaadinSession(VaadinService.java:857)
	at com.vaadin.flow.server.VaadinService.findOrCreateVaadinSession(VaadinService.java:832)
	at com.vaadin.flow.server.VaadinService.findVaadinSession(VaadinService.java:691)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1539)
```

--- a/flow-server/src/main/java/com/vaadin/flow/server/InitParameters.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/InitParameters.java
@@ -131,7 +131,9 @@ public class InitParameters implements Serializable {
 
     /**
      * Configuration parameter name for enabling session serialization in
-     * development.
+     * development. If enabled, all the session's associated
+     * {@link com.vaadin.flow.component.UI} instances will be serialized.
+     * Otherwise, it won't be serialized.
      *
      * @since
      */

--- a/flow-server/src/main/java/com/vaadin/flow/server/VaadinService.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/VaadinService.java
@@ -2175,10 +2175,6 @@ public abstract class VaadinService implements Serializable {
     protected VaadinSession readFromHttpSession(WrappedSession wrappedSession) {
         VaadinSession session = (VaadinSession) wrappedSession
                 .getAttribute(getSessionAttributeName());
-        if (session != null && session.isDeserializedAsEmpty()) {
-            wrappedSession.removeAttribute(getSessionAttributeName());
-            return null;
-        }
         return session;
     }
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/VaadinSession.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/VaadinSession.java
@@ -99,7 +99,7 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
     private LinkedList<RequestHandler> requestHandlers = new LinkedList<>();
 
     private int nextUIId = 0;
-    private Map<Integer, UI> uIs = new HashMap<>();
+    private transient Map<Integer, UI> uIs = new HashMap<>();
 
     protected WebBrowser browser = new WebBrowser();
 
@@ -141,8 +141,6 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
     private final Attributes attributes = new Attributes();
 
     private final StreamResourceRegistry resourceRegistry;
-
-    private transient boolean deserializedAsEmpty;
 
     /**
      * Creates a new VaadinSession tied to a VaadinService.
@@ -1031,10 +1029,6 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
     /**
      * Override default deserialization logic to account for transient
      * {@link #pendingAccessQueue}.
-     * <p>
-     * If a session was not serialized in writeObject, then only marks that this
-     * VaadinSession instance should not be used. The instance is removed by
-     * {@link VaadinService#readFromHttpSession(WrappedSession)}.
      *
      * @param stream
      *            the object to read
@@ -1043,20 +1037,13 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
      * @throws ClassNotFoundException
      *             if the class of the stream object could not be found
      */
+    @SuppressWarnings("unchecked")
     private void readObject(ObjectInputStream stream)
             throws IOException, ClassNotFoundException {
-        boolean wasSerialized = stream.readBoolean();
-        this.deserializedAsEmpty = !wasSerialized;
-        if (!wasSerialized) {
-            // even an empty unbound session still should support its API which
-            // requires a lock, so set a fake lock to be able to avoid
-            // unexpected issues on calling VaadinSession methods
-            lock = new ReentrantLock();
-            return;
-        }
         Map<Class<?>, CurrentInstance> old = CurrentInstance.setCurrent(this);
         try {
             stream.defaultReadObject();
+            uIs = (Map<Integer, UI>) stream.readObject();
             pendingAccessQueue = new ConcurrentLinkedQueue<>();
         } finally {
             CurrentInstance.restoreInstances(old);
@@ -1065,18 +1052,20 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
 
     private void writeObject(java.io.ObjectOutputStream stream)
             throws IOException {
-        boolean serialize = true;
+        boolean serializeUIs = true;
 
         ApplicationConfiguration appConfiguration = ApplicationConfiguration
                 .get(getService().getContext());
         if (!appConfiguration.isProductionMode()
                 && !appConfiguration.isDevModeSessionSerializationEnabled()) {
-            serialize = false;
+            serializeUIs = false;
         }
 
-        stream.writeBoolean(serialize);
-        if (serialize) {
-            stream.defaultWriteObject();
+        stream.defaultWriteObject();
+        if (serializeUIs) {
+            stream.writeObject(uIs);
+        } else {
+            stream.writeObject(new HashMap<>());
         }
     }
 
@@ -1118,13 +1107,5 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
      */
     public String getCsrfToken() {
         return csrfToken;
-    }
-
-    /**
-     * Checks if the session was deserialized without content and should be
-     * ignored and removed.
-     */
-    boolean isDeserializedAsEmpty() {
-        return deserializedAsEmpty;
     }
 }

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/ApplicationConfiguration.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/ApplicationConfiguration.java
@@ -91,6 +91,13 @@ public interface ApplicationConfiguration extends AbstractConfiguration {
     /**
      * Checks if development mode session serialization is enabled or not.
      * <p>
+     * Disabling session serialization means all its
+     * {@link com.vaadin.flow.component.UI} instances won't be serialized. This
+     * might be needed if one or more <code>UI</code>'s are not serializable
+     * and, thus, the whole http session might be discarded, making an
+     * authentication or other sensitive data stored in the session to get lost,
+     * which is not acceptable in most of the cases.
+     * <p>
      * By default session serialization is disabled in development mode.
      *
      * @return {@code true} if dev mode session serialization is enabled,

--- a/flow-server/src/test/java/com/vaadin/tests/server/SerializationTest.java
+++ b/flow-server/src/test/java/com/vaadin/tests/server/SerializationTest.java
@@ -6,20 +6,27 @@ import java.io.IOException;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
 import java.io.Serializable;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 
 import junit.framework.TestCase;
 import org.mockito.Mockito;
 
 import com.vaadin.flow.server.VaadinContext;
+import com.vaadin.flow.server.VaadinRequest;
 import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.VaadinServletService;
 import com.vaadin.flow.server.VaadinSession;
+import com.vaadin.flow.server.WrappedSession;
 import com.vaadin.flow.server.startup.ApplicationConfiguration;
+import com.vaadin.tests.util.MockUI;
 
 public class SerializationTest extends TestCase {
 
     public void testSerializeVaadinSession_accessQueueIsRecreated()
             throws Exception {
-        VaadinSession session = mockSession(true);
+        VaadinService vaadinService = new MockVaadinService(true);
+        VaadinSession session = new VaadinSession(vaadinService);
 
         session = serializeAndDeserialize(session);
 
@@ -28,33 +35,46 @@ public class SerializationTest extends TestCase {
                 session.getPendingAccessQueue());
     }
 
-    public void testSerializeVaadinSession_notProductionMode_disableDevModeSerialization_deserializedSessionHasLockInstance()
+    public void testSerializeVaadinSession_notProductionMode_disableDevModeSerialization_deserializedSessionHasNoUIs()
             throws Exception {
-        VaadinSession session = mockSession(false);
+        VaadinSession session = serializeAndDeserializeWithUI(false);
 
-        session = serializeAndDeserialize(session);
-
-        assertNotNull("Lock should be available after empty deserialization",
-                session.getLockInstance());
-        // self check : session is not really deserialized so the transient
-        // queue should be null
-        assertNull(session.getPendingAccessQueue());
+        assertNotNull("UIs should be available after empty deserialization",
+                session.getUIs());
+        assertTrue("UIs should be empty after empty deserialization",
+                session.getUIs().isEmpty());
     }
 
-    private VaadinSession mockSession(boolean productionMode) {
-        VaadinService service = Mockito.mock(VaadinService.class);
-        VaadinContext context = Mockito.mock(VaadinContext.class);
-        ApplicationConfiguration applicationConfiguration = Mockito
-                .mock(ApplicationConfiguration.class);
-        Mockito.when(service.getContext()).thenReturn(context);
-        Mockito.when(context.getAttribute(Mockito.any(), Mockito.any()))
-                .thenReturn(applicationConfiguration);
-        Mockito.when(applicationConfiguration.isProductionMode())
-                .thenReturn(productionMode);
-        Mockito.when(
-                applicationConfiguration.isDevModeSessionSerializationEnabled())
-                .thenReturn(false);
-        VaadinSession session = new VaadinSession(service);
+    public void testSerializeVaadinSession_notProductionMode_enableDevModeSerialization_deserializedSessionHasUI()
+            throws Exception {
+        VaadinSession session = serializeAndDeserializeWithUI(true);
+
+        assertNotNull("UIs should be available after empty deserialization",
+                session.getUIs());
+        assertEquals(
+                "UIs should contain a UI instance after empty deserialization",
+                1, session.getUIs().size());
+        assertEquals("Unexpected UI id after empty deserialization", 42,
+                session.getUIs().iterator().next().getUIId());
+    }
+
+    private static VaadinSession serializeAndDeserializeWithUI(
+            boolean serializeUI) throws IOException, ClassNotFoundException {
+        VaadinService vaadinService = new MockVaadinService(false, serializeUI);
+        VaadinSession session = new VaadinSession(vaadinService);
+        // This is done only for test purpose to init the session lock,
+        // should be called by Flow internally as soon as the session has
+        // been created.
+        session.refreshTransients(null, vaadinService);
+        MockUI ui = new MockUI(session);
+        ui.doInit(null, 42);
+        session.addUI(ui);
+
+        session = serializeAndDeserialize(session);
+        // This is done only for test purpose to refresh the session lock,
+        // should be called by Flow internally as soon as the session has
+        // been retrieved from http session.
+        session.refreshTransients(null, vaadinService);
         return session;
     }
 
@@ -112,4 +132,57 @@ public class SerializationTest extends TestCase {
             return dummyGetter;
         }
     }
+
+    private static class MockVaadinService extends VaadinServletService {
+
+        private final VaadinContext vaadinContext;
+        private final boolean productionMode;
+        private final boolean serialize;
+        private final Lock lock = new ReentrantLock();
+
+        {
+            lock.lock();
+        }
+
+        public MockVaadinService(boolean productionMode) {
+            super();
+            this.vaadinContext = Mockito.mock(VaadinContext.class);
+            this.productionMode = productionMode;
+            serialize = false;
+        }
+
+        public MockVaadinService(boolean productionMode, boolean serialize) {
+            super();
+            this.vaadinContext = Mockito.mock(VaadinContext.class);
+            this.productionMode = productionMode;
+            this.serialize = serialize;
+        }
+
+        @Override
+        public VaadinContext getContext() {
+            ApplicationConfiguration applicationConfiguration = Mockito
+                    .mock(ApplicationConfiguration.class);
+            Mockito.when(
+                    vaadinContext.getAttribute(Mockito.any(), Mockito.any()))
+                    .thenReturn(applicationConfiguration);
+            Mockito.when(applicationConfiguration.isProductionMode())
+                    .thenReturn(productionMode);
+            Mockito.when(applicationConfiguration
+                    .isDevModeSessionSerializationEnabled())
+                    .thenReturn(serialize);
+            return vaadinContext;
+        }
+
+        @Override
+        protected Lock getSessionLock(WrappedSession wrappedSession) {
+            return lock;
+        }
+
+        @Override
+        public String getMainDivId(VaadinSession session,
+                VaadinRequest request) {
+            return "main-div-id";
+        }
+    }
+
 }

